src/mlpack/tests/spill_tree_test.cpp

```diff
@@ -0,0 +1,247 @@
+#include <stack>
+
+using namespace mlpack;
+using namespace mlpack::tree;
+using namespace mlpack::metric;
+
+BOOST_AUTO_TEST_SUITE(SpillTreeTest);
+
+/**
+ * Test to make sure the tree contains the correct number of points after
+ * it is constructed. Also, it checks some invariants in the relation between
+ * parent and child nodes.
+ */
+BOOST_AUTO_TEST_CASE(SpillTreeConstructionCountTest)
+{
+  arma::mat dataset;
+  dataset.randu(3, 1000); // 1000 points in 3 dimensions.
+
+  typedef SPTree<EuclideanDistance, EmptyStatistic, arma::mat> TreeType;
+
+  // When overlapping buffer is 0, there shouldn't be repeated points.
+  TreeType tree1(dataset, 0);
+  TreeType tree2 = tree1;
+
+  BOOST_REQUIRE_EQUAL(tree1.NumDescendants(), 1000);
+  BOOST_REQUIRE_EQUAL(tree2.NumDescendants(), 1000);
+
+  // When overlapping buffer is greater than 0, it is possible to have repeated
+  // points. So, let's check node by node, that the number of descendants
+  // equals to the addition of the number of descendants of child nodes.
+  TreeType tree3(dataset, 0.5);
+
+  std::stack<TreeType*> nodes;
+  nodes.push(&tree3);
+  while (!nodes.empty())
+  {
+    TreeType* node = nodes.top();
+    nodes.pop();
+
+    size_t numDesc = node->NumPoints();
+
+    if (node->Left())
+    {
+      nodes.push(node->Left());
+      numDesc += node->Left()->NumDescendants();
+    }
+
+    if (node->Right())
+    {
+      nodes.push(node->Right());
+      numDesc += node->Right()->NumDescendants();
+    }
+
+    if (node->IsLeaf())
+      BOOST_REQUIRE_EQUAL(node->NumPoints(), node->NumDescendants());
+    else
+      BOOST_REQUIRE_EQUAL(node->NumPoints(), 0);
+
+    BOOST_REQUIRE_EQUAL(node->NumDescendants(), numDesc);
+  }
+}
+
+/**
+ * Test to check that parents and children are set correctly.
+ */
+BOOST_AUTO_TEST_CASE(SpillTreeConstructionParentTest)
+{
+  arma::mat dataset;
+  dataset.randu(3, 1000); // 1000 points in 3 dimensions.
+
+  typedef SPTree<EuclideanDistance, EmptyStatistic, arma::mat> TreeType;
+
+  TreeType tree(dataset, 0.5);
+
+  std::stack<TreeType*> nodes;
+  nodes.push(&tree);
+  while (!nodes.empty())
+  {
+    TreeType* node = nodes.top();
+    nodes.pop();
+
+    if (node->Left())
+    {
+      nodes.push(node->Left());
+      BOOST_REQUIRE_EQUAL(node, node->Left()->Parent());
+    }
+
+    if (node->Right())
+    {
+      nodes.push(node->Right());
+      BOOST_REQUIRE_EQUAL(node, node->Right()->Parent());
+    }
+  }
+}
+
+/**
+ * Auxiliary function to execute the same test for different flavours of Spill
+ * Trees.
+ */
+template<typename SpillType>
+void SpillTreeHyperplaneTestAux()
+{
+  arma::mat dataset;
+  dataset.randu(3, 1000); // 1000 points in 3 dimensions.
+
+  for (size_t cases = 0; cases < 3; cases++)
+  {
+    double tau;
+    switch (cases)
+    {
+      case 0:
+        tau = 0;
+        break;
+      case 1:
+        tau = 0.05;
+        break;
+      case 2:
+        tau = 0.1;
+        break;
+    }
```

Comment:
Done in https://github.com/mlpack/mlpack/pull/747/commits/f989164147b33ba87192c311fff19622cc59f714